### PR TITLE
feat: Introduce clusters

### PR DESCRIPTION
**Pull request by lynnagara**
Introduce the cluster class. The cluster provides a reader
and clickhouse connections shared by all storages that are
located on the cluster.

This PR does not yet change the settings format in any way and hardcodes
a single global cluster (essentially the same as what we had before).
This will be added in future.

### CODE REVIEW

File: snuba/datasets/storage.py
Comment (fpacifici) [inline snuba/datasets/storage.py:26]: Why does the storage need to receive the cluster if the `get_cluster` method takes a storage key. This does not seem right since the cluster is not a structural, hardcoded property of the storage (the StorageSet would be). The mapping storageset-cluster should come from the config since the existence of a Cluster is only in config.

File: snuba/datasets/groups.py
Comment (fpacifici) [inline snuba/datasets/groups.py:35]: Does this actually type check? It seems to me the parent class takes in a Cluster.

File: snuba/datasets/storages/factory.py
Comment (fpacifici) [inline snuba/datasets/storages/factory.py:21]: ?

File: snuba/web/query.py
Comment (fpacifici) [inline snuba/web/query.py:109]: I don't think this is going to work for joined queries. We need to assume multiple storages can be used in the same query. I think there are 2 issues here:
1) easier to solve: I am not comfortable in leaking the storage class at this point from the query plan builder. The plan buidler returns a query plan, and that is supposed to know how to run the query (it can run a query across multiple storages which are at that point hidden from this code). You can let the plan builder provide a cluster to the query plan and get the cluster from the queryplan. See an example of how it could work (I need to wrap that PR) here https://github.com/getsentry/snuba/pull/830/files#diff-87784c0acdd074c79dca3bec3bb2d403R16.
2) I would not even leak the reader here. Why don't you just let the QueryPlanBuilder provide the cluster or the StorageSet to the QueryPlanExecutionStrategy that will pass the cluster to the QueryRunner  which will resolve the reader and do the rest ? I am suggesting this because it should make the information flow easier to understand. We initialize the query runner only with the data the QueryPlan cannot know about (like the timer). The rest (that is the outcomes of the decisions made by QueryPlanBuilder) flows directly from QueryPlanBuilder to QueryExecutionStrategy. But I do not have a strong opinion on this last point. As long as we do not depend on a single storage and we do not leak the storage class here I am ok.

**Review by fpacifici — COMMENTED** [review]
First round of high level comments. I think the most important issue is that the lack of StorageSet (which should be a very simple abstraction is making it hard to grasp how this could work with queries that span multiple storages. 
I would consider adding that here even if it makes the PR a little longer.

File: snuba/cli/bulk_load.py
Comment (tkaemming) [inline snuba/cli/bulk_load.py:28]: Good catch.

File: snuba/snapshots/loaders/__init__.py
Comment (tkaemming) [inline snuba/snapshots/loaders/__init__.py:17]: I don't think we anticipated this as one of the impacted areas of this change, but this seems similar to the reader in that we'll want to find some sort of way to generalize this over time to break the dependency on the ClickHouse client directly.

Both the writer and the ClickHouse client here are derived the same writable storage when they are passed in from the command line tool. Should the writable storage be the thing that is passed here instead?

File: snuba/clusters/__init__.py
Comment (tkaemming) [inline snuba/clusters/__init__.py:47]: This works here, but I think trying to consolidate these resources in `snuba.environment` will be helpful over the long term because it enables us to eventually consolidate all of the different object instances that need to have some sort of lifecycle control.

There is a tech debt ticket for this, and [some more details here](https://github.com/getsentry/snuba/pull/772#discussion_r377381301), but I guess as of right now this is just a personal preference and not anything we've really discussed the overall merits and direction of… maybe we should?

File: snuba/clusters/__init__.py
Comment (tkaemming) [inline snuba/clusters/__init__.py:7]: I think it would make sense to split this into two classes (both concrete):

1. A `Cluster` that exposes `get_reader`.
2. A `ClickhouseCluster` that subclasses `Cluster` and exposes the additional methods that return the ClickHouse clients.

I think in an idealized model we'd be able to only depend on `Cluster` and all of the ClickHouse implementation details would be abstracted away, but it in the mean time this would allow us to understand what is still dependent on the ClickHouse-specific functionality.

File: snuba/clusters/__init__.py
Comment (tkaemming) [reply snuba/clusters/__init__.py:54]: I agree with this as well.

I also think that the storage set abstraction is sufficiently well contained that it'd be doable to implement that change on it's own, and merge it in here afterwards to break up this change?

File: snuba/datasets/dataset.py
Comment (tkaemming) [reply snuba/datasets/dataset.py:123]: I haven't fully thought through this, but if it's possible to consolidate these two methods (`get_writable_storage` and `get_table_writer`) as well as `enforce_table_writer`, I'd be elated.

File: snuba/web/db_query.py
Comment (tkaemming) [inline snuba/web/db_query.py:62]: Doesn't this need a bound type variable?

**Review by tkaemming — COMMENTED** [review]
Some scattershot comments inline from my first pass over this.

I think that already being familiar with the ideas here has helped me a lot as a reviewer to understand what is going on here without having to reverse engineer the intent of the code in front of me, I hope it made implementation similarly focused, this seems like a fast turnaround for such a major abstraction and overall a really good start.

File: snuba/clusters/storage_sets.py
Comment (fpacifici) [inline snuba/clusters/storage_sets.py:11]: I think we can improve types here.
If you create a type alias for StorageSet (that maps to string) and a type alias for the storage_key (that maps to string), then we can properly qualify storage set names and storage keys in methods signatures instead of using strings and the type checker can enforce that.

But beyond that, referring to storages by storage_key here is calling for integrity issues (a typo here would show up only in production). Why don't you define storage keys either through constants or through an Enum ? So the type checker can enforce we only reference valid ones. Also you could reference the instance of the storage, but that would bake in the assumption that storages are singleton at runtime which we do not need.

File: snuba/datasets/storage.py
Comment (fpacifici) [inline snuba/datasets/storage.py:41]: I think this would be confusing. It would make people think that storages are directly assigned to clusters while they are assigned only through StorageSet. I think you could either document that relationship here (in details) considering this is the method in the abstract class, or just return the storage set that just makes it clear.

File: snuba/datasets/storages/errors.py
Comment (fpacifici) [inline snuba/datasets/storages/errors.py:152]: Same as I mentioned above, these should be either constants or an enum.

File: snuba/web/query.py
Comment (fpacifici) [reply snuba/web/query.py:109]: still applies

File: snuba/clusters/storage_sets.py
Comment (fpacifici) [inline snuba/clusters/storage_sets.py:4]: I think this relation storageset -> storage is going in the wrong direction.If you make the storageset define a set of storages you are building an m..n relationship . It is only by convention at runtime that you are enforcing that is is actually a 1..n.
If you let the storage instance point to the storage_set you do not have this problem. There structurally it would be impossible to miss a storage or to assign a storage to multiple storage sets, it also comes with the bonus that you do not need the storage_key anymore.

Is there any use case where you cannot go that way ?

**Review by fpacifici — CHANGES_REQUESTED** [review]
Thanks for adding the storageset. I think we should be more restrictive on the typing there. See comments inline. There are quite a few guarantees of correctness we can enforce with the type checker instead of at runtime.
Also, please see my comments around leaking the storage to web/query. They still apply.

File: snuba/clusters/cluster.py
Comment (markstory) [inline snuba/clusters/cluster.py:22]:
```suggestion
    In the future, clusters will also be responsible for coordinating commands that
```

File: snuba/clusters/storage_sets.py
Comment (lynnagara) [reply snuba/clusters/storage_sets.py]: There was not really a reason except that I thought it'd be slightly easier to see the list of storage sets required to be registered. I've changed the structure of this to go the other way.

File: snuba/clusters/__init__.py
Comment (lynnagara) [reply snuba/clusters/__init__.py]: Which writer would be moving here? TableWriter? I thought that one will not be shared for the cluster but specific to each storage/table.

File: snuba/datasets/groups.py
Comment (lynnagara) [reply snuba/datasets/groups.py:36]: Was testing a couple of options. Ended up changing the parent class to take storage key.

File: snuba/web/query.py
Comment (lynnagara) [reply snuba/web/query.py]: Thanks, I've been looking for a better solution for this. I changed the builder to pass the cluster to the query plan execution strategy.

File: snuba/clusters/cluster.py
Comment (lynnagara) [reply snuba/clusters/cluster.py:24]: I can try to reword this, there are 4 examples listed on the line below (bootstrap, migrate, cleanup, optimize)

File: snuba/clusters/__init__.py
Comment (lynnagara) [reply snuba/clusters/__init__.py]: Currently joins (well, the 1 join we have) are registered as a storage on the same clusters as events so retrieving the cluster works the same as any other storage. In future, it seems like joins won't be a storage but will contain 2 other storages. As long as those 2 storages being joined are on the same cluster (which is enforced by the storage set), this should still work. I wanted to avoid exposing the storage set everywhere unnecessarily, when what we usually want is to go from storage to cluster.

File: snuba/datasets/dataset.py
Comment (lynnagara) [reply snuba/datasets/dataset.py]: Done outside of this PR, will remove enforce_table_writer after updating the consumer and replacer in https://github.com/getsentry/snuba/pull/875 and https://github.com/getsentry/snuba/pull/861

File: snuba/snapshots/loaders/__init__.py
Comment (lynnagara) [reply snuba/snapshots/loaders/__init__.py:17]: Yeah, I think passing the writable storage here makes sense (at least for now).

Comment (lynnagara) [reply snuba/snapshots/loaders/__init__.py:17]: I did not yet find a good way to do this without introducing a circular dependency where the writer depends on the higher level storage.

File: snuba/clusters/storage_sets.py
Comment (fpacifici) [inline snuba/clusters/storage_sets.py:9]: groupassignees is supposed to be joined with events so it should be in the same storage set.

File: snuba/clusters/storage_sets.py
Comment (fpacifici) [inline snuba/clusters/storage_sets.py:29]: It seems this relationship is going the wrong way.
It is still a mapping between StorageSet and a set of Storages.
This means it could be an 1:n or an m:n.
I think you should have a get_storage_set method in the storage. That forces a 1:n relationship between storageset and storage. Like we do for foreign keys in databases.

Then getting the cluster from storage becomes trivial and you may not even need the storageKey anymore.

File: snuba/web/query.py
Comment (fpacifici) [inline snuba/web/query.py:111]: way better, thanks.

File: snuba/clusters/storage_sets.py
Comment (fpacifici) [inline snuba/clusters/storage_sets.py:7]: Now this is the only definition of what a storage set is. It needs a docstring ecplaining what a storage set it.

File: snuba/datasets/storages/__init__.py
Comment (fpacifici) [inline snuba/datasets/storages/__init__.py:4]: Please add a docstring to explain what a StorageKey is and which constraints it has (like that it has to be unique per storage class)

File: snuba/web/db_query.py
Comment (fpacifici) [reply snuba/web/db_query.py:62]: it does.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Better, I think we are close. Still two remarks on the way we assign storage set to storages and how we expose the cluster from the execution strategy

File: snuba/clusters/__init__.py
Comment (fpacifici) [reply snuba/clusters/__init__.py]: ping

File: snuba/clusters/__init__.py
Comment (fpacifici) [reply snuba/clusters/__init__.py]: ping

File: snuba/web/query.py
Comment (lynnagara) [reply snuba/web/query.py:123]: So, reader is passed to runner from the execution strategy rather than `_run_query_pipeline`? That sounds good to me and lets us remove the `get_cluster()` which only existed for that reason.

Re: 1 and 2, is there a reason why runner should be changed to take the cluster as well instead of just the reader like before though?

Comment (fpacifici) [reply snuba/web/query.py:123]: Runner is not taking the cluster not the reader now. It is just the partially applied function that gets the reader. I am talking about this https://github.com/getsentry/snuba/blob/master/snuba/datasets/plans/query_plan.py#L12

And I would pass cluster instead of reader because it is a more abstract concept, and doing that does not add complexity.

Comment (lynnagara) [reply snuba/web/query.py:123]: Sorry, the last part was supposed to read `_format_storage_query_and_run` is taking the reader and passing it through to `raw_query`, not runner.

File: snuba/clusters/storage_sets.py
Comment (lynnagara) [reply snuba/clusters/storage_sets.py:33]: I guess I was trying to add the storage set mapping without exposing this anywhere outside the clusters module. The fact that storages are mapped to clusters only through storage sets is a detail that is mostly hidden in here. Kind of subjective but I also find it easier to glance at the mappings in this way rather than open the ~9 files for each storage.

You do need to remember to add it to this list, but it will yell and you and you won't be able to run the API or any command and nearly every test will fail if not.

File: snuba/environment.py
Comment (fpacifici) [inline snuba/environment.py:41]: There is a comment related with this here:
https://github.com/getsentry/snuba/pull/860#discussion_r402032120

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:52]: This comment is still valid:
https://github.com/getsentry/snuba/pull/860#discussion_r402033694

**Review by fpacifici — CHANGES_REQUESTED** [review]
I think there are still a couple of pending comments and we should revisit how we deal with the bulk loader.

File: snuba/clusters/__init__.py
Comment (lynnagara) [reply snuba/clusters/__init__.py]: Agree we should decide on a consistent approach, can we hold off for now? Most other resources are not handled in that way yet. There's a bit more going on in here now and I'd prefer to keep it centralised in clusters until we make a decision.

File: snuba/clusters/__init__.py
Comment (lynnagara) [reply snuba/clusters/__init__.py]: Updated

File: snuba/cli/consumer.py
Comment (lynnagara) [reply snuba/cli/consumer.py]: Yeah, thanks

File: snuba/snapshots/loaders/__init__.py
Comment (lynnagara) [reply snuba/snapshots/loaders/__init__.py:17]: It's intentional in that I think many other changes need to happen here and I don't want to tackle it within the scope of this PR. 

- There are quite a number of places including this one where settings.CLICKHOUSE_HOST and friends are still being referenced, these all need to be removed of getting the information from the cluster. (optimize, cleanup, etc)
- I think we should probably just pass a storage into the bulk load command and remove the ability to pass a dataset and destination table. The dest_table comes from the storage.
- Also update confirm_load to do the same
- It depends on moving the writer (assuming we want to move it into storage set), which is not in scope of this PR

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:83]: you are using the set of registered stored set twice. You can assign it to a variable so you do not need to recreate it.
Also you do not need to create a list of StorageKeys first and then turn it into a set. `set(s for s in StorageSetKey)` would work.

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:91]: I think you should be able to build this with a dict comprehension.

File: snuba/datasets/storage.py
Comment (fpacifici) [inline snuba/datasets/storage.py:33]: nit: just get_storage_set. storage_key works for the storage, where the key is only a reference to the storage. the storage_set is entirely identified by the enum value instead, no need to call it a key.

File: snuba/snapshots/loaders/__init__.py
Comment (fpacifici) [reply snuba/snapshots/loaders/__init__.py:17]: I am ok holding off on this. Could you just mention with a TODO that this interface is going to change and in which direction ?

**Review by fpacifici — APPROVED** [review]
I think this is solid, thanks for doing this and for applying all the comments.

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:10]: Wait, I think this hierarchy is inverted.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Sorry I think I missed one comment last night.

File: snuba/clusters/cluster.py
Comment (fpacifici) [reply snuba/clusters/cluster.py:10]: Cluster should be the parent class and only expose the abstract concepts like reader and storage set, while the subclass (clickhouse should be used only when you need the clickhouse specific concept). This allows us to drive the design in a direction that only depends on the abstract class (and as a consequence being independent on the implementation).
Was there a reason to make the parent class, clickhouse specific?

File: snuba/datasets/storage.py
Comment (lynnagara) [reply snuba/datasets/storage.py:33]: Well, it's more to match the name of the class which is still called `StorageSetKey` (despite the fact there is no `StorageSet` class)

File: snuba/clusters/cluster.py
Comment (lynnagara) [reply snuba/clusters/cluster.py]: This was a mistake, inverted these classes now.

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:36]: well, indeed, making the Reader clickhouse specific means this does not work anymore because the Cluster interface becomes clickhouse specific.
You can make this class generic itself and based on the abstract generic reader class and materialize the parameter type in the clickhouse implementation.

```
class Cluster(ABC, Generic[TQuery]):
     def get_reader(self) -> Reader[TQuery]

class ClickhouseCluster(Cluster[ClickhouseQuery]):
     def get_reader(self) -> Reader[ClickhouseQuery]

```

I think the callsites should still type check because the everywhere this is used, we already reference the ClickhouseCluster explicitly

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:56]: nit: putting the parent class initialization at the beginning makes it teasier to see the correctness of the method.

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:53]: Out of curiosity,  did mypy complain without the explicit typehint here ? I feel it should not.

File: snuba/datasets/storage.py
Comment (fpacifici) [reply snuba/datasets/storage.py:33]: ok

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:71]: Since now we have clickhouse clusters you may want to call this CLICKHOUSE_CLUSTERS, but won't block for this. Up to you.

**Review by fpacifici — COMMENTED** [review]
I think this is fine. Just a couple of remarks, if you want to implement them. I know Ted wanted to review this as well, so I will wait for that before accepting.

File: snuba/clusters/cluster.py
Comment (lynnagara) [reply snuba/clusters/cluster.py:36]: Ah, thank you.

File: snuba/clusters/cluster.py
Comment (lynnagara) [reply snuba/clusters/cluster.py]: Yeah, you're right it wasn't needed.

File: snuba/clusters/cluster.py
Comment (tkaemming) [inline snuba/clusters/cluster.py:33]: For consistency with: https://github.com/getsentry/snuba/pull/860/files#r413246184

```suggestion
    def get_storage_set_keys(self) -> Set[StorageSetKey]:
        return {StorageSetKey(storage_set) for storage_set in self.__storage_sets}
```

**Review by tkaemming — COMMENTED** [review]
This looks good and should be very helpful moving forward. I noted one minor thing inline that was a bit hard to understand at a glance.

File: snuba/clusters/cluster.py
Comment (lynnagara) [reply snuba/clusters/cluster.py:33]: Thank you

**Review by fpacifici — APPROVED** [review]
thanks for doing this.

**Pull request merged**
